Add color-scheme article (#2481)
GoogleChrome/web.dev: src/site/content/en/blog/color-scheme/index.md

---

> title: The `color-scheme` CSS property and the corresponding meta tag

Author: ```suggestion
title: Improved dark mode with the `color-scheme` CSS property and the corresponding meta tag
```

At the same time, this feature prevents browsers from applying any transformations on their own.

### The user-agent stylesheet

Before I continue, let me briefly describe what a user-agent stylesheet is.
Most of the times, you can think of the word *user agent* (UA)
as a fancy way to say *browser*.
The UA stylesheet determines the default look and feel of a page.
As the name suggests, a UA stylesheet is something that depends on the UA in question.
You can have a look at
[Chrome's](https://chromium.googlesource.com/chromium/blink/+/master/Source/core/css/html.css)
(and Chromium's) UA stylesheet and compare it to
[Firefox's](https://dxr.mozilla.org/mozilla-central/source/layout/style/res/html.css) or
[Safari's](https://trac.webkit.org/browser/trunk/Source/WebCore/css/html.css) (and WebKit's).
Typically, UA stylesheets agree on the majority of things.
For example, they all make links blue, general text black, and background color white,
but there are also important (and sometimes annoying) differences,
for instance, how they style form controls.

Have a closer look at
[WebKit's UA stylesheet](https://trac.webkit.org/browser/trunk/Source/WebCore/css/html.css)

> and what it does regarding dark mode (do a full text search for "dark" in the code).

Author: ```suggestion
and what it does regarding dark mode.
(Do a full text search for "dark" in the stylesheet.)
```

---
title: The `color-scheme` CSS property and the corresponding meta tag
subhead: |

>   The `color-scheme` CSS property and the corresponding meta tag

Author: It does.

  allow developers to opt in their pages to theme-specific defaults of the user-agent stylesheet.
authors:
  - thomassteiner
date: 2020-03-27
hero: hero.jpg
alt: Pigeons on a wall with a sharp black and white contrast in the background.
description: |

>   The color-scheme CSS property and the corresponding meta tag

Author: This is for the description meta tag, so here it doesn't show.

  allow developers to opt their pages in to theme-specific defaults of the user-agent stylesheet,
  such as, for example, form controls, scroll bars, as well as CSS system colors.
  At the same time, this feature prevents browsers from applying any transformations on their own.
tags:
  - post # post is a required tag for the article to show up in the blog.
  - css
  - dark-mode
  - dark-theme
  - prefers-color-scheme
  - color-scheme
draft: true
---
## Background

### The `prefers-color-scheme` user preference media feature

The
[`prefers-color-scheme`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-color-scheme)
user preference media feature gives developers full control over their pages' appearances.

> If you're unfamiliar with it read my article

Author: ```suggestion
If you are unfamiliar with it read my article
```

Most of the times, you can think of the word *user agent* (UA)
as a fancy way to say *browser*.
The UA stylesheet determines the default look and feel of a page.
As the name suggests, a UA stylesheet is something that depends on the UA in question.
You can have a look at
[Chrome's](https://chromium.googlesource.com/chromium/blink/+/master/Source/core/css/html.css)
(and Chromium's) UA stylesheet and compare it to
[Firefox's](https://dxr.mozilla.org/mozilla-central/source/layout/style/res/html.css) or
[Safari's](https://trac.webkit.org/browser/trunk/Source/WebCore/css/html.css) (and WebKit's).
Typically, UA stylesheets agree on the majority of things.
For example, they all make links blue, general text black, and background color white,
but there are also important (and sometimes annoying) differences,
for instance, how they style form controls.

Have a closer look at
[WebKit's UA stylesheet](https://trac.webkit.org/browser/trunk/Source/WebCore/css/html.css)
and what it does regarding dark mode.
(Do a full text search for "dark" in the stylesheet.)
The default provided by the stylesheet changes based on whether dark mode is on or off.

> To illustrate this, here is one such CSS rule (using the
> [`:matches`](https://css-tricks.com/almanac/selectors/m/matches/)
> pseudo class and WebKit-internal variables like `-apple-system-control-background`,
> as well as the WebKit-internal preprocessor directive `#if defined`):

Author: ```suggestion
To illustrate this, here is one such CSS rule using the
[`:matches`](https://css-tricks.com/almanac/selectors/m/matches/)
pseudo class and WebKit-internal variables like `-apple-system-control-background`,
as well as the WebKit-internal preprocessor directive `#if defined`:
```

      HAVE_OS_DARK_MODE_SUPPORT
    color: text;
    background-color: -apple-system-control-background;
  #else
    background-color: white;
  #endif
  /* snip */
}
```

You will notice some non-standard values for the `color` and `background-color` properties above.
Neither `text` nor `-apple-system-control-background` are valid CSS colors.
They are WebKit-internal *semantic* colors.

Turns out, CSS has standardized semantic system colors.
They are specified in
[CSS Color Module Level&nbsp;4](https://drafts.csswg.org/css-color/#css-system-colors).
For example,
[`Canvas`](https://drafts.csswg.org/css-color/#valdef-system-color-canvas)
(not to be confused with the `<canvas>` tag)
is for the background of application content or documents,
whereas
[`CanvasText`](https://drafts.csswg.org/css-color/#valdef-system-color-canvastext)
is for text in application content or documents.
The two go together and should not be used in isolation.

UA stylesheets can make use of their own proprietary, or the standardized semantic system colors,
to determine how HTML elements should be rendered by default.
If the operating system is set to dark mode or uses a dark theme,

> `CanvasText` (or `text` respectively) would be conditionally set to whitish,

Author: `text` is the WebKit-internal way to mean `CanvasText`.

> `CanvasText` (or `text` respectively) would be conditionally set to white,

Author: ```suggestion
`CanvasText` (or `text`) would be conditionally set to white,
```

> `CanvasText` (or `text` respectively) would be conditionally set to whitish,

Author: ```suggestion
`CanvasText` (or `text` respectively) would be conditionally set to white,
```

> and `Canvas` (or `-apple-system-control-background`) would be set to blackish.

Author: ```suggestion
and `Canvas` (or `-apple-system-control-background`) would be set to black.
```

The UA stylesheet then assigns the following CSS only once, and covers both light and dark mode.

```css
/**
  Not actual UA stylesheet code.
  For illustrative purposes only.
*/
body {
  color: CanvasText;
  background-color: Canvas
}
```

## The `color-scheme` CSS property

The [CSS Color Adjustment Module Level&nbsp;1](https://drafts.csswg.org/css-color-adjust/)
specification introduces a model and controls

> over automatic color adjustment by the user agent to handle user preferences,

Author: ```suggestion
over automatic color adjustment by the user agent
with the objective of handling user preferences
```

> such as dark mode, contrast adjustment, or specific desired color schemes.

Author: It's straight from the spec, including the wrong comma. I've fixed it.

eventually result in the same behavior, I always recommend specifying the color scheme
via the meta tag, so the browser can adopt to the preferred scheme faster.

{% Aside 'warning' %}
  While for absolute baseline pages no additional CSS rules are necessary,
  in the general case you should always combine `color-scheme` with `prefers-color-scheme`.

  For example, the proprietary WebKit CSS color `-webkit-link`, used by WebKit and Chrome
  for the classic link blue `rgb(0,0,238)`,
  has an insufficient contrast ratio of 2.23:1 on a black background and
  [fails](https://webaim.org/resources/contrastchecker/?fcolor=0000EE&bcolor=000000)
  both the WCAG&nbsp;AA as well as the WCAG&nbsp;AAA
  [requirements](https://www.w3.org/WAI/WCAG21/Understanding/conformance#levels).
{% endAside %}

## Interplay with `prefers-color-scheme`

The interplay of the `color-scheme` CSS property and the corresponding meta tag
with the `prefers-color-scheme` user preference media feature may seem confusing at first.
In fact, they play together really well.
The most important thing to understand is that `color-scheme`
exclusively determines the default appearance,

> whereas `prefers-color-scheme` also determines the stylable appearance.

Reviewer: ```suggestion
whereas `prefers-color-scheme` determines the stylable appearance.
```

To make this clearer, let's assume the following page:

```html
<head>
  <meta name="color-scheme" content="dark light">
  <style>
    fieldset {
      background-color: gainsboro;
    }
    @media (prefers-color-scheme: dark) {
      fieldset {
        background-color: darkslategray;
      }
    }
  </style>
</head>
<body>
  <p>
    Lorem ipsum dolor sit amet, legere ancillae ne vis.
  </p>
  <form>
    <fieldset>
      <legend>Lorem ipsum</legend>
      <button type="button">Lorem ipsum</button>
    </fieldset>
  </form>
</body>
```

The inline CSS code on the page

> sets the `<fieldset>`'s `background-color` to `gainsboro` in the general case,

Reviewer: ```suggestion
sets the `<fieldset>` element's `background-color` to `gainsboro` in the general case,
```

and to `darkslategray` if the user prefers a `dark` color scheme
according to the `prefers-color-scheme` user preference media feature.

Via the `<meta name="color-scheme" content="dark light">` element,
the page tells the browser that it supports a dark and a light theme,
with a preference for a dark theme.

Depending on whether the operating system is set to dark or light mode,
the whole page appears light on dark, or vice versa, based on the user-agent stylesheet.
There is *no* additional developer-provided CSS involved to change the paragraph text
or the background color of the page.

> Note how the `<fieldset>`'s `background-color` changes

Reviewer: ```suggestion
Note how the `<fieldset>` element's `background-color` changes
```

> based on whether dark mode is enabled or not, following the rules

Reviewer: ```suggestion
based on whether dark mode is enabled, following the rules
```

in the developer-provided inline stylesheet on the page.
It is either `gainsboro` or `darkslategray`.

<figure>
  <img src="light-styles.png" width="3440" height=" 1386">
  <figcaption>
    <strong>Light mode:</strong> Styles specified by the developer and the user-agent.
    The text is black and the background is white as per the user-agent stylesheet.

>     The <code>&lt;fieldset&gt;</code>'s <code>background-color</code> is <code>gainsboro</code>

Reviewer: ```suggestion
    The <code>&lt;fieldset&gt;</code> element's <code>background-color</code> is <code>gainsboro</code>
```

    as per the inlined developer stylesheet.
  </figcaption>
</figure>

<figure>
  <img src="dark-styles.png" width="3440" height=" 1386">
  <figcaption>
    <strong>Dark mode:</strong> Styles specified by the developer and the user-agent.
    The text is white and the background is black as per the user-agent stylesheet.

>     The <code>&lt;fieldset&gt;</code>'s <code>background-color</code> is <code>darkslategray</code>

Reviewer: ```suggestion
    The <code>&lt;fieldset&gt;</code> element's <code>background-color</code> is <code>darkslategray</code>
```

    as per the inlined developer stylesheet.
  </figcaption>
</figure>

> The `<button>`'s appearance is controlled by the user-agent stylesheet.

Reviewer: ```suggestion
The `<button>` element's appearance is controlled by the user-agent stylesheet.
```

Its `color` is set to the
[`ButtonText`](https://drafts.csswg.org/css-color/#valdef-system-color-buttontext)
system color, and its `background-color` and the four `border-color`s are set to the system color
[`ButtonFace`](https://drafts.csswg.org/css-color/#valdef-system-color-buttonface)

<figure>
  <img src="light-buttonface.png" width="3440" height=" 1386">
  <figcaption>
    <strong>Light mode:</strong> The <code>background-color</code> and the various

>     <code>border-color</code>s are set the the <a href="https://drafts.csswg.org/css-color/#valdef-system-color-buttonface">ButtonFace</a>

Reviewer: ```suggestion
    <code>border-color</code>s are set to the <a href="https://drafts.csswg.org/css-color/#valdef-system-color-buttonface">ButtonFace</a>
```

    system color.
  </figcaption>
</figure>

> Now note how the `<button>`'s `border-color` changes.

Reviewer: ```suggestion
Now note how the `<button>` element's `border-color` changes.
```

The *computed* value for the `border-top-color` and the `border-bottom-color`
switches from `rgba(0, 0, 0, 0.847)` (blackish) to `rgba(255, 255, 255, 0.847)` (whitish),
since the user-agent updates `ButtonFace` dynamically based on the color scheme.

> The same applies for the `<button>`'s `color`

Reviewer: ```suggestion
The same applies for the `<button>` element's `color`
```

that is set to the corresponding system color `ButtonText`.

<figure>
  <img src="light-computed.png" width="3440" height=" 1386">
  <figcaption>
    <strong>Light mode:</strong> The computed values of the <code>border-top-color</code>
    and the <code>border-bottom-color</code> that are both set to <code>ButtonFace</code>
    in the user-agent stylesheet are now <code>rgba(0, 0, 0, 0.847)</code>.
  </figcaption>
</figure>

<figure>
  <img src="dark-computed.png" width="3440" height=" 1386">
  <figcaption>
    <strong>Dark mode:</strong> The computed values of the <code>border-top-color</code>
    and the <code>border-bottom-color</code> that are both set to <code>ButtonFace</code>
    in the user-agent stylesheet are now <code>rgba(255, 255, 255, 0.847)</code>.
  </figcaption>
</figure>

## Demo

You can see the effects of `color-scheme` applied to a large number of HTML elements
in the demo embedded below, or
[directly on Glitch](https://color-scheme-demo.glitch.me/).
Remember that you need to switch your operating system to dark mode to enjoy this demo.

> The demo deliberately uses the WCAG&nbsp;AA and WCAG&nbsp;AAA

Author: ```suggestion
The demo deliberately shows the WCAG&nbsp;AA and WCAG&nbsp;AAA
```

[violation](https://webaim.org/resources/contrastchecker/?fcolor=0000EE&bcolor=000000)
with the link colors that was mentioned in the

> [warning above](#using-color-scheme-in-practice).

Reviewer: ```suggestion
The demo deliberately uses the WCAG&nbsp;AA and WCAG&nbsp;AAA
[violation](https://webaim.org/resources/contrastchecker/?fcolor=0000EE&bcolor=000000)
with the link colors mentioned in the
[warning above](#using-color-scheme-in-practice).
```

{% Aside 'note' %}
  The `color-scheme` CSS property can be used on both the `:root` level,
  as well as on an individual per-element level.
{% endAside %}

On the `:root` element, rendering with a color scheme
additionally must affect the surface color of the canvas (that is, the global background color),
the initial value of the `color` property, and the used values of the system colors,
and should also affect the viewport's scroll bars.

```css
/*
  The page supports both dark and light color schemes,
  and the page author prefers dark.
*/
:root {
  color-scheme: dark light;
}
```

## The `color-scheme` meta tag

Honoring the `color-scheme` CSS property requires the CSS to be first downloaded
(if it is referenced via `<link rel="stylesheet">`) and to be parsed.
To aid user agents in rendering the page background with the desired color scheme *immediately*,
a `color-scheme` value can also be provided in a
[`<meta name="color-scheme">`](https://drafts.csswg.org/css-color-adjust/#color-scheme-meta)
element.

```html
<!--
  The page supports both dark and light color schemes,
  and the page author prefers dark.
-->
<meta name="color-scheme" content="dark light">
```

> ## Using color-scheme in practice

Author: ```suggestion
## Combining color-scheme and prefers-color-scheme
```

```css
/*
  The page supports both dark and light color schemes,
  and the page author prefers dark.
*/
:root {
  color-scheme: dark light;
}
```

## The `color-scheme` meta tag

Honoring the `color-scheme` CSS property requires the CSS to be first downloaded
(if it is referenced via `<link rel="stylesheet">`) and to be parsed.
To aid user agents in rendering the page background with the desired color scheme *immediately*,
a `color-scheme` value can also be provided in a
[`<meta name="color-scheme">`](https://drafts.csswg.org/css-color-adjust/#color-scheme-meta)
element.

```html
<!--
  The page supports both dark and light color schemes,
  and the page author prefers dark.
-->
<meta name="color-scheme" content="dark light">
```

## Using color-scheme in practice

Since both the meta tag and the CSS property (if applied to the `:root` element)
eventually result in the same behavior, I always recommend specifying the color scheme
via the meta tag, so the browser can adopt to the preferred scheme faster.

While for absolute baseline pages no additional CSS rules are necessary,
in the general case you should always combine `color-scheme` with `prefers-color-scheme`.

> 
> For example, the proprietary WebKit CSS color `-webkit-link`, used by WebKit and Chrome

Author: ```suggestion
For example, the proprietary WebKit CSS color `-webkit-link`, used by WebKit and Chrome
```

the initial value of the `color` property, and the used values of the system colors,
and should also affect the viewport's scroll bars.

```css
/*
  The page supports both dark and light color schemes,
  and the page author prefers dark.
*/
:root {
  color-scheme: dark light;
}
```

## The `color-scheme` meta tag

Honoring the `color-scheme` CSS property requires the CSS to be first downloaded
(if it is referenced via `<link rel="stylesheet">`) and to be parsed.
To aid user agents in rendering the page background with the desired color scheme *immediately*,
a `color-scheme` value can also be provided in a
[`<meta name="color-scheme">`](https://drafts.csswg.org/css-color-adjust/#color-scheme-meta)
element.

```html
<!--
  The page supports both dark and light color schemes,
  and the page author prefers dark.
-->
<meta name="color-scheme" content="dark light">
```

## Using color-scheme in practice

Since both the meta tag and the CSS property (if applied to the `:root` element)
eventually result in the same behavior, I always recommend specifying the color scheme
via the meta tag, so the browser can adopt to the preferred scheme faster.

> {% Aside 'warning' %}
>   While for absolute baseline pages no additional CSS rules are necessary,
>   in the general case you should always combine `color-scheme` with `prefers-color-scheme`.
> 
>   For example, the proprietary WebKit CSS color `-webkit-link`, used by WebKit and Chrome
>   for the classic link blue `rgb(0,0,238)`,
>   has an insufficient contrast ratio of 2.23:1 on a black background and
>   [fails](https://webaim.org/resources/contrastchecker/?fcolor=0000EE&bcolor=000000)
>   both the WCAG&nbsp;AA as well as the WCAG&nbsp;AAA
>   [requirements](https://www.w3.org/WAI/WCAG21/Understanding/conformance#levels).
> 
>   I have opened bugs for [Chrome](https://crbug.com/1066811),
>   [WebKit](https://bugs.webkit.org/show_bug.cgi?id=209851), and
>   [Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1626560)
>   as well as a [meta issue in the HTML Standard](https://github.com/whatwg/html/issues/5426)
>   to get this fixed.
> {% endAside %}

Author: ```suggestion
While for absolute baseline pages no additional CSS rules are necessary,
in the general case you should always combine `color-scheme` with `prefers-color-scheme`.

For example, the proprietary WebKit CSS color `-webkit-link`, used by WebKit and Chrome
for the classic link blue `rgb(0,0,238)`,
has an insufficient contrast ratio of 2.23:1 on a black background and
[fails](https://webaim.org/resources/contrastchecker/?fcolor=0000EE&bcolor=000000)
both the WCAG&nbsp;AA as well as the WCAG&nbsp;AAA
[requirements](https://www.w3.org/WAI/WCAG21/Understanding/conformance#levels).

I have opened bugs for [Chrome](https://crbug.com/1066811),
[WebKit](https://bugs.webkit.org/show_bug.cgi?id=209851), and
[Firefox](https://bugzilla.mozilla.org/show_bug.cgi?id=1626560)
as well as a [meta issue in the HTML Standard](https://github.com/whatwg/html/issues/5426)
to get this fixed.
```
